data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/DefaultSqlPreparedQuery.java

```diff
@@ -222,10 +227,14 @@ public static Sort enhanceCursoredSort(Sort sort, boolean isBackwards, Persisten
         // sorting on the rows. Therefore, we make sure id is present in it.
         List<Order> orders = new ArrayList<>(sort.getOrderBy());
         for (PersistentProperty idProperty: persistentEntity.getIdentityProperties()) {
-            String name = idProperty.getName();
-            if (orders.stream().noneMatch(o -> o.getProperty().equals(name))) {
-                orders.add(Order.asc(name));
-            }
+            PersistentEntityUtils.traversePersistentProperties(idProperty, (associations, property) -> {
+                String prefix = String.join(".", associations.stream().map(Association::getName).collect(Collectors.toList()));
+                String propertyName = property.getName();
+                String name = StringUtils.isEmpty(prefix) ? propertyName : prefix + "." + propertyName;
```

Comment:
Should we use the naming stratefy here somehow?

Reply:
We need to build `shipmentId.country` etc. here and not with `_` like strategy has methods for so maybe we can leave it like this for just this case.